src/rust/engine/ui/src/display.rs

```diff
@@ -0,0 +1,307 @@
+extern crate rand;
+extern crate termion;
+extern crate unicode_segmentation;
+
+use std::collections::{HashMap, VecDeque};
+use std::io::Write;
+use std::io::{stdout, Result, Stdout};
+use std::thread;
+use std::time::Duration;
+
+use termion::cursor::DetectCursorPos;
+use termion::raw::IntoRawMode;
+use termion::raw::RawTerminal;
+use termion::{clear, color, cursor};
+use unicode_segmentation::UnicodeSegmentation;
+
+enum Console {
+  Terminal(RawTerminal<Stdout>),
+  Pipe(Stdout),
+}
+
+pub struct EngineDisplay {
+  sigil: char,
+  divider: String,
+  poll_interval_ms: Duration,
+  padding: String,
+  terminal: Console,
+  action_map: HashMap<String, String>,
+  logs: VecDeque<String>,
+  running: bool,
+  is_tty: bool,
+  cursor_start: (u16, u16),
+  terminal_size: (u16, u16),
+}
+
+// TODO: Prescribe a threading/polling strategy for callers - or implement a built-in one.
```

Comment:
Before landing this, please take a look at which TODOs deserve more fleshed out github issues vs just inline expansion for a reader who might not have all of the context.

Reply:
will do.